### DISCUSSION TITLE
Constant high CPU usage of immich-machine-learning

### DISCUSSION THREAD

**Discussion by femto-code** [immich-app/immich#3160] [Q&A] [answered]
I am getting a rather high, constant CPU usage causing the entire system to be unstable (high temperature, delays). Considering the docker overview, it is coming from the machine learning module. I haven't uploaded any photos that would explain it to work that much though.

My `.env` and `docker-compose.yml` are the default ones, I only adjusted the path for the upload dir. I would consider turning off the feature but I would really like to use it on the other hand. Does anyone know what might cause an issue here? Or is the machine not powerful enough?

![image](https://github.com/immich-app/immich/assets/59444046/af086d63-88f6-4556-a545-65a81430b1dc)

The following log messages come repeatedly, as if the server process is being restarted over and over...
```
2023-07-08 20:54:17 INFO:     Started server process [1]
2023-07-08 20:54:17 INFO:     Waiting for application startup.
2023-07-08 20:54:18 Could not find image processor class in the image processor config or the model config. Loading based on pattern matching with the model's feature extractor configuration.
2023-07-08 20:54:22 /opt/venv/lib/python3.11/site-packages/onnxruntime/capi/onnxruntime_inference_collection.py:65: UserWarning: Specified provider 'CUDAExecutionProvider' is not in available provider names.Available providers: 'CPUExecutionProvider'
2023-07-08 20:54:22   warnings.warn(
2023-07-08 20:53:59 Applied providers: ['CPUExecutionProvider'], with options: {'CPUExecutionProvider': {}}
2023-07-08 20:54:22 Applied providers: ['CPUExecutionProvider'], with options: {'CPUExecutionProvider': {}}
```

My system:
```
OS: Kubuntu 23.04 x86_64
Kernel: 6.2.0-24-generic
CPU: Intel i7-2670QM (8) @ 3.100GHz
GPU: AMD ATI Radeon HD 6630M/6650M/6750M/7670M/7690M 
GPU: Intel 2nd Generation Core Processor Family
```

**Comment by mertalev** [comment]
It seems to be constantly restarting due to insufficient memory. The logs show that it hasn't loaded all models and seems stuck while loading one of the facial recognition models. You can try setting `MACHINE_LEARNING_EAGER_STARTUP=false` in your `.env` so models are loaded when actively used instead of all of them being loaded at startup. 

However, when uploading images all ML jobs run afterwards by default, so for this to help you also need to pause all ML jobs except for one so only one model needs to be loaded at a time.

**Reply by femto-code** [reply]
Thanks! Okay, I assume increasing the memory available to the docker image is a solution? I can see that more RAM would be available but it is not using it for some reason.

**Comment by alextran1502** [comment, marked as answer]
You would need 4-6 GB of memory for Immich stack, because machine-learning would load the model into memories when processing and then later on it would release those RAM

**Reply by femto-code** [reply]
Thanks for your help! Where can I see how much RAM got allocated for the Immich stack? Can I change that amount somewhere?

**Reply by darrylgibbs** [reply]
@alextran1502 I'm guessing this is a "it depends" type of question, but I fired up a new Immich stack last night (6c / 8GB RAM - i5 10500t for reference) in a Proxmox LXC Container.  All works well but the ML and MS containers for me have been pinned on 200%+ utilization overnight for a 200gb library.  Physical CPU usage is 80% and RAM is 2.5/8GB.  Is this normal behaviour?  Or is a VM better suited?

**Reply by alextran1502** [reply]
@darrylgibbs if the jobs are still running it would be normal behavior. I am running LXC as well with docker-compose

**Reply by darrylgibbs** [reply]
@alextran1502 awesome, I'll let it run.  Thank you sir!
